crates/core/tedge/src/cli/certificate/cli.rs

```diff
@@ -193,16 +196,187 @@ pub enum UploadCertCli {
+            Some(config_id) => Ok(config_id.into()),
+        }
+        Some(input_id) if config_id.is_none() => Ok(input_id),
+        Some(input_id) if input_id == config_id.unwrap() => Ok(input_id),
+        Some(input_id) => match cloud.as_ref().map(<_>::into) {
+            None => Err(MismatchedDeviceId {
+                input_id,
+                config_id: config_id.unwrap().into(),
+                writable_key,
+            }),
+            Some(tedge_config::Cloud::C8y(profile)) => {
+                let key = profile.map(|name| name.to_string());
+                let c8y_dto = dto.c8y.try_get(key.as_deref(), "c8y")?;
+                if c8y_dto.device.id.is_some() {
+                    Err(MismatchedDeviceId {
+                        input_id,
+                        config_id: config_id.unwrap().into(),
+                        writable_key,
+                    })
+                } else {
+                    Ok(input_id)
+                }
+            }
+            Some(tedge_config::Cloud::Az(profile)) => {
+                let key = profile.map(|name| name.to_string());
+                let az_dto = dto.az.try_get(key.as_deref(), "az")?;
+                if az_dto.device.id.is_some() {
+                    Err(MismatchedDeviceId {
+                        input_id,
+                        config_id: config_id.unwrap().into(),
+                        writable_key,
+                    })
+                } else {
+                    Ok(input_id)
+                }
+            }
+            Some(tedge_config::Cloud::Aws(profile)) => {
+                let key = profile.map(|name| name.to_string());
+                let aws_dto = dto.c8y.try_get(key.as_deref(), "aws")?;
+                if aws_dto.device.id.is_some() {
+                    Err(MismatchedDeviceId {
+                        input_id,
+                        config_id: config_id.unwrap().into(),
+                        writable_key,
+                    })
+                } else {
+                    Ok(input_id)
+                }
+            }
+        },
+    }
+}
+
+pub(crate) fn get_writable_key(cloud: &Option<Cloud>) -> Result<WritableKey, anyhow::Error> {
+    let key = match cloud {
+        None => WritableKey::DeviceId,
+        Some(cloud) => {
+            let key = match cloud {
+                Cloud::C8y(_) => WritableKey::C8yDeviceId(None),
+                Cloud::Azure(_) => WritableKey::AzDeviceId(None),
+                Cloud::Aws(_) => WritableKey::AwsDeviceId(None),
+            };
+            let profile = cloud.profile_name().cloned();
+            crate::try_with_profile!(key, profile)
+        }
+    };
+    Ok(key)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use tedge_test_utils::fs::TempTedgeDir;
+    use test_case::test_case;
+
+    #[test_case("test", None, None, toml::toml!{
+    [device]
+        id = "test"
+    })]
```

Comment:
Formatting suggestion:
```suggestion
    #[test_case(
        "test", 
        None, 
        None, 
        toml::toml!{
            [device]
            id = "test"
        }
    )]
```
Same for the rest of the test cases as well.

Reply:
Which formatter are you using and forces this? My IDE doesn't, it's a pain to do all manually :sweat_smile:

Reply:
That was just manually formatted as `cargo fmt` doesn't touch stuff defined inside macros (`test_case` in this case). I found that laset few test cases really hard to read, hence suggested this.

Reply:
Addressed in https://github.com/thin-edge/thin-edge.io/commit/99620aa1a5cc761ec794b7097e36008398a9360d